docs/advanced_topics/customisation/rich_text_internals.rst

```diff
@@ -0,0 +1,205 @@
+ * Enforcing a subset of HTML helps to keep presentational markup out of the database, making the site more maintainable, and making it easier to repurpose site content (including, potentially, producing non-HTML output such as `DocBook <https://docbook.org/>`_).
+ * Elements such as page links and images need to preserve metadata such as the page or image ID, which is not present in the final HTML representation.
+
+This requires the rich text content to go through a number of validation and conversion steps; both between the editor interface and the version stored in the database, and from the database representation to the final rendered HTML.
+
+For this reason, extending Wagtail's rich text handling to support a new element is more involved than simply saying (for example) "enable the ``<blockquote>`` element", since various components of Wagtail - both client and server-side - need to agree on how to handle that feature, including how it should be exposed in the editor interface, how it should be represented within the database, and (if appropriate) how it should be translated when rendered on the front-end.
+
+The components involved in Wagtail's rich text handling are described below.
+
+
+Data format
+-----------
+
+Rich text data (as handled by :ref:`RichTextField <rich-text>`, and ``RichTextBlock`` within :doc:`StreamField </topics/streamfield>`) is stored in the database in a format that is similar, but not identical, to HTML. For example, a link to a page might be stored as:
+
+.. code-block:: html
+
+    <p><a linktype="page" id="3">Contact us</a> for more information.</p>
+
+Here, the ``linktype`` attribute identifies a rule that shall be used to rewrite the tag. When rendered on a template through the ``|richtext`` filter (see :ref:`rich-text-filter`), this is converted into valid HTML:
+
+.. code-block:: html
+
+    <p><a href="/contact-us/">Contact us</a> for more information.</p>
+
+In the case of ``RichTextBlock``, the block's value is a ``RichText`` object which performs this conversion automatically when rendered as a string, so the ``|richtext`` filter is not necessary.
+
+Likewise, an image inside rich text content might be stored as:
+
+.. code-block:: html
+
+    <embed embedtype="image" id="10" alt="A pied wagtail" format="left" />
+
+which is converted into an ``img`` element when rendered:
+
+.. code-block:: html
+
+    <img alt="A pied wagtail" class="richtext-image left" height="294" src="/media/images/pied-wagtail.width-500_ENyKffb.jpg" width="500">
+
+Again, the ``embedtype`` attribute identifies a rule that shall be used to rewrite the tag. All tags other than ``<a linktype="...">`` and ``<embed embedtype="..." />`` are left unchanged in the converted HTML.
+
+A number of additional constraints apply to ``<a linktype="...">`` and ``<embed embedtype="..." />`` tags, to allow the conversion to be performed efficiently via string replacement:
+
+ * The tag name and attributes must be lower-case;
```

Comment:
I love semicolons but I think they look a bit old-school in ULs :)

Reply:
Fair point! Removed in 96d81d78f1f178ed42fd90803ba61e6ba4f603ed.